File: app/src/main/java/com/dantsu/thermalprinter/MainActivity.java
package com.dantsu.thermalprinter;


import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.lifecycle.Observer;
import androidx.recyclerview.widget.RecyclerView;

import android.Manifest;
import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.AlertDialog;
import android.app.ProgressDialog;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.view.inputmethod.InputMethodManager;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageButton;

import com.dantsu.escposprinter.connection.DeviceConnection;
import com.dantsu.escposprinter.connection.bluetooth.BluetoothConnection;
import com.dantsu.escposprinter.connection.bluetooth.BluetoothPrintersConnections;
import com.dantsu.thermalprinter.async.AsyncBluetoothEscPosPrint;
import com.dantsu.thermalprinter.async.AsyncEscPosPrint;
import com.dantsu.thermalprinter.async.AsyncEscPosPrinter;
import com.dantsu.thermalprinter.data.Comment;
import com.dantsu.thermalprinter.viewmodel.MainViewModel;
import com.google.gson.Gson;


import java.net.URISyntaxException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Date;
import java.util.List;

import io.socket.client.IO;
import io.socket.client.Socket;
import io.socket.emitter.Emitter;

public class MainActivity extends AppCompatActivity {


    private Socket mSocket;
    {
        try {
            mSocket = IO.socket("https://www.ttliveprint.com/");
        } catch (URISyntaxException ignored) {}
    }



    MainViewModel mainViewModel = new MainViewModel();

    private final List<Comment> comments = new ArrayList<>();
    private RecyclerView recyclerView;

    private ItemAdapter adapter;

    private EditText editText;

    private Button btnSelectDevice;

    private  ImageButton btnScrollDown;
    private Button btnConnectServer;
    private CustomLinearLayoutManager linearLayoutManager;

    private ProgressDialog progressDialog;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);


        setContentView(R.layout.activity_main);
        editText = this.findViewById(R.id.edtUniqueId);
        btnSelectDevice = this.findViewById(R.id.button_bluetooth_browse);
        btnScrollDown = this.findViewById(R.id.btn_scroll_down);
        btnConnectServer = this.findViewById(R.id.connect_to_server);
        btnSelectDevice.setOnClickListener(view -> browseBluetoothDevice());
        btnConnectServer.setOnClickListener(view -> connectToServer());
        btnScrollDown.setOnClickListener(view -> scrollDown());
        recyclerView = findViewById(R.id.recyclerView);
        linearLayoutManager = new CustomLinearLayoutManager(this);
        recyclerView.setLayoutManager(linearLayoutManager);
        adapter = new ItemAdapter(comments);
        adapter.setOnItemClickListener(this::printBluetooth);
        recyclerView.setAdapter(adapter);
        progressDialog = new ProgressDialog(this);
        progressDialog.setMessage("Connecting...");
        progressDialog.setCancelable(false);

        recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                if (dy < 0) {
                    // User is scrolling up
                    mainViewModel.setIsOnBottom(false);
                    btnScrollDown.setVisibility(View.VISIBLE); // Show the button
                }else {
                    if(linearLayoutManager.findLastVisibleItemPosition()== adapter.getItemCount()-1){
                        mainViewModel.setIsOnBottom(true);
                        btnScrollDown.setVisibility(View.INVISIBLE);
                    }
                }
            }
        });
                mainViewModel.getIsOnBottom().observe(this, (Observer) o -> {
                    if(o instanceof Boolean){
                        if(o == Boolean.TRUE){
                            btnScrollDown.setVisibility(View.INVISIBLE);
                        } else{
                            btnScrollDown.setVisibility(View.VISIBLE);
                        }
                    }
                });

    }
    private void connectToServer(){
        progressDialog.show();
        hideKeyboard(this);
        String uniqueId = "huyentuixach1";
        if(!editText.getText().toString().equals("")){
            uniqueId = editText.getText().toString();
        }

        mSocket.connect();
        mSocket.on("connect",onNewMessage);
        mSocket.emit("setUniqueId",uniqueId);
        mSocket.on("chat",onChat);
        mSocket.on("disconnect", onDisconnected);
        mSocket.on("streamEnd", onStreamEnded);
    }
    Emitter.Listener onChat = args -> runOnUiThread(() -> {
        progressDialog.hide();
        try{
            editText.setText("");
            editText.setVisibility(View.GONE);
            btnConnectServer.setClickable(false);
            btnSelectDevice.setClickable(false);
            Comment comment = new Gson().fromJson(args[0].toString(),Comment.class);
            comments.add(comment);
            adapter.notifyItemInserted(comments.size()-1);
            if(Boolean.TRUE.equals(mainViewModel.getIsOnBottom().getValue())){
                recyclerView.smoothScrollToPosition(comments.size()-1);
                btnScrollDown.setVisibility(View.INVISIBLE);
            }

        } catch (Exception e){
            e.printStackTrace();
        }
    });

    @SuppressLint("NotifyDataSetChanged")
    Emitter.Listener onStreamEnded = args -> runOnUiThread(()->{
        try{
            mSocket.emit("disconnect", "Hello" );
            mSocket.disconnect();
            mSocket.close();
            btnConnectServer.setClickable(true);
            btnSelectDevice.setClickable(true);
            editText.setVisibility(View.VISIBLE);
            comments.clear();
            adapter.notifyDataSetChanged();
        } catch (Exception e){
            e.printStackTrace();
        }
    });


    private final Emitter.Listener onNewMessage = args -> runOnUiThread(() -> Log.d("Main Activity", Arrays.toString(args)));


    private final Emitter.Listener onDisconnected = args -> runOnUiThread(()->{
        btnSelectDevice.setClickable(true);
        btnConnectServer.setClickable(true);
        editText.setVisibility(View.VISIBLE);
        mSocket.off("new message", onNewMessage);
        mSocket.close();
    });


    private void scrollDown(){
        mainViewModel.onScrollDown();
        recyclerView.smoothScrollToPosition(comments.size()-1);
    }

    public static void hideKeyboard(Activity activity) {
        InputMethodManager imm = (InputMethodManager) activity.getSystemService(Activity.INPUT_METHOD_SERVICE);
        //Find the currently focused view, so we can grab the correct window token from it.
        View view = activity.getCurrentFocus();
        //If no view currently has focus, create a new one, just so we can grab a window token from it
        if (view == null) {
            view = new View(activity);
        }
        imm.hideSoftInputFromWindow(view.getWindowToken(), 0);
    }
    @Override
    protected void onStop() {

        super.onStop();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        mSocket.disconnect();
        mSocket.off("new message", onNewMessage);
    }
    /*==============================================================================================
    ======================================BLUETOOTH PART============================================
    ==============================================================================================*/

    public interface OnBluetoothPermissionsGranted {
        void onPermissionsGranted();
    }

    public static final int PERMISSION_BLUETOOTH = 1;
    public static final int PERMISSION_BLUETOOTH_ADMIN = 2;
    public static final int PERMISSION_BLUETOOTH_CONNECT = 3;
    public static final int PERMISSION_BLUETOOTH_SCAN = 4;

    public OnBluetoothPermissionsGranted onBluetoothPermissionsGranted;

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            switch (requestCode) {
                case MainActivity.PERMISSION_BLUETOOTH:
                case MainActivity.PERMISSION_BLUETOOTH_ADMIN:
                case MainActivity.PERMISSION_BLUETOOTH_CONNECT:
                case MainActivity.PERMISSION_BLUETOOTH_SCAN:
                    this.checkBluetoothPermissions(this.onBluetoothPermissionsGranted);
                    break;
            }
        }
    }

    public void checkBluetoothPermissions(OnBluetoothPermissionsGranted onBluetoothPermissionsGranted) {
        this.onBluetoothPermissionsGranted = onBluetoothPermissionsGranted;
        if (android.os.Build.VERSION.SDK_INT < android.os.Build.VERSION_CODES.S && ContextCompat.checkSelfPermission(this, Manifest.permission.BLUETOOTH) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.BLUETOOTH}, MainActivity.PERMISSION_BLUETOOTH);
        } else if (android.os.Build.VERSION.SDK_INT < android.os.Build.VERSION_CODES.S && ContextCompat.checkSelfPermission(this, Manifest.permission.BLUETOOTH_ADMIN) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.BLUETOOTH_ADMIN}, MainActivity.PERMISSION_BLUETOOTH_ADMIN);
        } else if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.S && ContextCompat.checkSelfPermission(this, Manifest.permission.BLUETOOTH_CONNECT) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.BLUETOOTH_CONNECT}, MainActivity.PERMISSION_BLUETOOTH_CONNECT);
        } else if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.S && ContextCompat.checkSelfPermission(this, Manifest.permission.BLUETOOTH_SCAN) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.BLUETOOTH_SCAN}, MainActivity.PERMISSION_BLUETOOTH_SCAN);
        } else {
            this.onBluetoothPermissionsGranted.onPermissionsGranted();
        }
    }

    private BluetoothConnection selectedDevice;

    public void browseBluetoothDevice() {
        this.checkBluetoothPermissions(() -> {
            final BluetoothConnection[] bluetoothDevicesList = (new BluetoothPrintersConnections()).getList();

            if (bluetoothDevicesList != null) {
                final String[] items = new String[bluetoothDevicesList.length + 1];
                items[0] = "Default printer";
                int i = 0;
                for (BluetoothConnection device : bluetoothDevicesList) {
                    if (ActivityCompat.checkSelfPermission(this, Manifest.permission.BLUETOOTH_CONNECT) != PackageManager.PERMISSION_GRANTED) {
                        return;
                    }
                    items[++i] = device.getDevice().getName();

                }

                AlertDialog.Builder alertDialog = new AlertDialog.Builder(MainActivity.this);
                alertDialog.setTitle("Bluetooth printer selection");
                alertDialog.setItems(
                    items,
                    (dialogInterface, i1) -> {
                        int index = i1 - 1;
                        if (index == -1) {
                            selectedDevice = null;
                        } else {
                            selectedDevice = bluetoothDevicesList[index];
                        }
                        Button button = findViewById(R.id.button_bluetooth_browse);
                        button.setText(items[i1]);
                    }
                );

                AlertDialog alert = alertDialog.create();
                alert.setCanceledOnTouchOutside(false);
                alert.show();
            }
        });

    }

    public void printBluetooth(int position) {
        this.checkBluetoothPermissions(() -> new AsyncBluetoothEscPosPrint(
            this,
            new AsyncEscPosPrint.OnPrintFinished() {
                @Override
                public void onError(AsyncEscPosPrinter asyncEscPosPrinter, int codeException) {
                    Log.e("MainActivity", "AsyncEscPosPrint.OnPrintFinished : An error occurred !");
                }

                @Override
                public void onSuccess(AsyncEscPosPrinter asyncEscPosPrinter) {
                    Log.i("MainActivity", "AsyncEscPosPrint.OnPrintFinished : Print is finished !");
                }
            }
        )
            .execute(this.getAsyncEscPosPrinter(selectedDevice, position)));
    }


    /**
     * Asynchronous printing
     */
    @SuppressLint("SimpleDateFormat")
    public AsyncEscPosPrinter getAsyncEscPosPrinter(DeviceConnection printerConnection, int position) {
        SimpleDateFormat format = new SimpleDateFormat("'on' yyyy-MM-dd 'at' HH:mm:ss");
        AsyncEscPosPrinter printer = new AsyncEscPosPrinter(printerConnection, 203, 58f, 47);
        return printer.addTextToPrint(
                    "[L]\n" +
                    "[L]<u>HUYEN TUI XACH</u>\n" +
                    "[L]\n" +
                    "[L]<u type='double'>" + format.format(new Date()) + "</u>\n" +
                    "[L]======================\n" +
                    "[L]<b>Comment:</b>"+ comments.get(position).comment+"\n" +
                    "[L]<b>Nick name:</b>"+comments.get(position).nickname+"\n" +
                    "[L]<b>Unique Id:</b>"+comments.get(position).uniqueId+"\n" +
                    "[L]======================\n" +
                    "[L]<qrcode size='20'>"+ "https://www.tiktok.com/@"+comments.get(position).uniqueId+"</qrcode>\n"+ "[L]\n" + "[L]\n"
        );
    }
}

File: app/src/main/java/com/dantsu/thermalprinter/MainActivity.kt
package com.dantsu.thermalprinter

import android.Manifest
import android.annotation.SuppressLint
import android.app.Activity
import android.app.AlertDialog
import android.app.ProgressDialog
import android.content.DialogInterface
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.View
import android.view.inputmethod.InputMethodManager
import android.widget.Button
import android.widget.EditText
import android.widget.ImageButton
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import com.dantsu.escposprinter.connection.DeviceConnection
import com.dantsu.escposprinter.connection.bluetooth.BluetoothConnection
import com.dantsu.escposprinter.connection.bluetooth.BluetoothPrintersConnections
import com.dantsu.thermalprinter.async.AsyncBluetoothEscPosPrint
import com.dantsu.thermalprinter.async.AsyncEscPosPrint.OnPrintFinished
import com.dantsu.thermalprinter.async.AsyncEscPosPrinter
import com.dantsu.thermalprinter.data.Comment
import com.dantsu.thermalprinter.viewmodel.MainViewModel
import com.google.gson.Gson
import io.socket.client.IO
import io.socket.client.Socket
import io.socket.emitter.Emitter
import java.net.URISyntaxException
import java.text.SimpleDateFormat
import java.util.Arrays
import java.util.Date

class MainActivity : AppCompatActivity() {
    private var mSocket: Socket? = null

    init {
        try {
            mSocket = IO.socket("https://www.ttliveprint.com/")
        } catch (ignored: URISyntaxException) {
        }
    }

    var mainViewModel = MainViewModel()
    private val comments: MutableList<Comment> = ArrayList()
    private lateinit var recyclerView: RecyclerView
    private var adapter: ItemAdapter? = null
    private var editText: EditText? = null
    private lateinit var btnSelectDevice: Button
    private lateinit var btnScrollDown: ImageButton
    private lateinit var btnConnectServer :Button
    private var linearLayoutManager: CustomLinearLayoutManager? = null
    private var progressDialog: ProgressDialog? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        editText = findViewById(R.id.edtUniqueId)
        btnSelectDevice = findViewById(R.id.button_bluetooth_browse)
        btnScrollDown = findViewById(R.id.btn_scroll_down)
        btnConnectServer = findViewById(R.id.connect_to_server)
        btnSelectDevice.setOnClickListener { browseBluetoothDevice() }
        btnConnectServer.setOnClickListener { connectToServer() }
        btnScrollDown.setOnClickListener { scrollDown() }
        recyclerView = findViewById(R.id.recyclerView)
        linearLayoutManager = CustomLinearLayoutManager(this)
        recyclerView.layoutManager = linearLayoutManager
        adapter = ItemAdapter(comments)
        adapter!!.setOnItemClickListener(object:ItemAdapter.OnItemClickListener{
            override fun onItemClick(position: Int) {
                printBluetooth(position)
            }
        })
        recyclerView.adapter = adapter
        progressDialog = ProgressDialog(this)
        progressDialog!!.setMessage("Connecting...")
        progressDialog!!.setCancelable(false)
        recyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                if (dy < 0) {
                    // User is scrolling up
                    mainViewModel.setIsOnBottom(false)
                    btnScrollDown.visibility = View.VISIBLE // Show the button
                } else {
                    if (linearLayoutManager!!.findLastVisibleItemPosition() == adapter!!.itemCount - 1) {
                        mainViewModel.setIsOnBottom(true)
                        btnScrollDown.visibility = View.INVISIBLE
                    }
                }
            }
        })
        mainViewModel.isOnBottom.observe(this
        ) {
            if (it == true) {
                btnScrollDown.visibility = View.INVISIBLE
            } else {
                btnScrollDown.visibility = View.VISIBLE
            }
        }
    }

    private fun connectToServer() {
        progressDialog!!.show()
        hideKeyboard(this)
        var uniqueId = "huyentuixach"
        if (editText!!.text.toString() != "") {
            uniqueId = editText!!.text.toString()
        }
        mSocket!!.connect()
        mSocket!!.on("connect", onNewMessage)
        mSocket!!.emit("setUniqueId", uniqueId)
        mSocket!!.on("chat", onChat)
        mSocket!!.on("disconnect", onDisconnected)
        mSocket!!.on("streamEnd", onStreamEnded)
    }

    private var onChat = Emitter.Listener { args: Array<Any> ->
        runOnUiThread {
            progressDialog!!.hide()
            try {
                editText!!.setText("")
                editText!!.visibility = View.GONE
                btnConnectServer.isClickable = false
                btnSelectDevice.isClickable = false
                val comment = Gson().fromJson(args[0].toString(), Comment::class.java)
                comments.add(comment)
                adapter!!.notifyItemInserted(comments.size - 1)
                if (java.lang.Boolean.TRUE == mainViewModel.isOnBottom.value) {
                    recyclerView.smoothScrollToPosition(comments.size - 1)
                    btnScrollDown.visibility = View.INVISIBLE
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    @SuppressLint("NotifyDataSetChanged")
    var onStreamEnded = Emitter.Listener { _: Array<Any?>? ->
        runOnUiThread {
            try {
                mSocket!!.emit("disconnect", "Hello")
                mSocket!!.disconnect()
                mSocket!!.close()
                btnConnectServer.isClickable = true
                btnSelectDevice.isClickable = true
                editText!!.visibility = View.VISIBLE
                comments.clear()
                adapter!!.notifyDataSetChanged()
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }
    private val onNewMessage = Emitter.Listener { args: Array<Any?>? ->
        runOnUiThread {
            Log.d(
                "Main Activity",
                Arrays.toString(args)
            )
        }
    }
    private val onDisconnected = Emitter.Listener { _: Array<Any?>? ->
        runOnUiThread {
            btnSelectDevice.isClickable = true
            btnConnectServer.isClickable = true
            editText!!.visibility = View.VISIBLE
            mSocket!!.off("new message", onNewMessage)
            mSocket!!.close()
        }
    }

    private fun scrollDown() {
        mainViewModel.onScrollDown()
        recyclerView.smoothScrollToPosition(comments.size - 1)
    }

    public override fun onDestroy() {
        super.onDestroy()
        mSocket!!.disconnect()
        mSocket!!.off("new message", onNewMessage)
    }

    /*==============================================================================================
    ======================================BLUETOOTH PART============================================
    ==============================================================================================*/
    interface OnBluetoothPermissionsGranted {
        fun onPermissionsGranted()
    }

    var onBluetoothPermissionsGranted: OnBluetoothPermissionsGranted? = null
    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (grantResults.isNotEmpty() && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            when (requestCode) {
                PERMISSION_BLUETOOTH, PERMISSION_BLUETOOTH_ADMIN, PERMISSION_BLUETOOTH_CONNECT, PERMISSION_BLUETOOTH_SCAN -> checkBluetoothPermissions(
                    onBluetoothPermissionsGranted
                )
            }
        }
    }

    private fun checkBluetoothPermissions(onBluetoothPermissionsGranted: OnBluetoothPermissionsGranted?) {
        this.onBluetoothPermissionsGranted = onBluetoothPermissionsGranted
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.S && ContextCompat.checkSelfPermission(
                this,
                Manifest.permission.BLUETOOTH
            ) != PackageManager.PERMISSION_GRANTED
        ) {
            ActivityCompat.requestPermissions(
                this,
                arrayOf(Manifest.permission.BLUETOOTH),
                PERMISSION_BLUETOOTH
            )
        } else if (Build.VERSION.SDK_INT < Build.VERSION_CODES.S && ContextCompat.checkSelfPermission(
                this,
                Manifest.permission.BLUETOOTH_ADMIN
            ) != PackageManager.PERMISSION_GRANTED
        ) {
            ActivityCompat.requestPermissions(
                this,
                arrayOf(Manifest.permission.BLUETOOTH_ADMIN),
                PERMISSION_BLUETOOTH_ADMIN
            )
        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S && ContextCompat.checkSelfPermission(
                this,
                Manifest.permission.BLUETOOTH_CONNECT
            ) != PackageManager.PERMISSION_GRANTED
        ) {
            ActivityCompat.requestPermissions(
                this,
                arrayOf(Manifest.permission.BLUETOOTH_CONNECT),
                PERMISSION_BLUETOOTH_CONNECT
            )
        } else if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S && ContextCompat.checkSelfPermission(
                this,
                Manifest.permission.BLUETOOTH_SCAN
            ) != PackageManager.PERMISSION_GRANTED
        ) {
            ActivityCompat.requestPermissions(
                this,
                arrayOf(Manifest.permission.BLUETOOTH_SCAN),
                PERMISSION_BLUETOOTH_SCAN
            )
        } else {
            this.onBluetoothPermissionsGranted!!.onPermissionsGranted()
        }
    }

    private var selectedDevice: BluetoothConnection? = null
    private fun browseBluetoothDevice() {
        checkBluetoothPermissions(object:OnBluetoothPermissionsGranted {
            override fun onPermissionsGranted() {
                val bluetoothDevicesList = BluetoothPrintersConnections().list
                if (bluetoothDevicesList != null) {
                    val items = arrayOfNulls<String>(bluetoothDevicesList.size + 1)
                    items[0] = "Default printer"
                    for ((i, device) in bluetoothDevicesList.withIndex()) {
                        if (ActivityCompat.checkSelfPermission(
                                this@MainActivity,
                                Manifest.permission.BLUETOOTH_CONNECT
                            ) != PackageManager.PERMISSION_GRANTED
                        ) {
                            return
                        }
                        items[i + 1] = device.device.name
                    }
                    val alertDialog = AlertDialog.Builder(this@MainActivity)
                    alertDialog.setTitle("Bluetooth printer selection")
                    alertDialog.setItems(
                        items
                    ) { _: DialogInterface?, i1: Int ->
                        val index = i1 - 1
                        selectedDevice = if (index == -1) {
                            null
                        } else {
                            bluetoothDevicesList[index]
                        }
                        val button = findViewById<Button>(R.id.button_bluetooth_browse)
                        button.text = items[i1]
                    }
                    val alert = alertDialog.create()
                    alert.setCanceledOnTouchOutside(false)
                    alert.show()
                }
            }
        })
    }

    private fun printBluetooth(position: Int) {
        checkBluetoothPermissions(object:OnBluetoothPermissionsGranted{
            override fun onPermissionsGranted() {
                    AsyncBluetoothEscPosPrint(
                        this@MainActivity,
                        object : OnPrintFinished() {
                            override fun onError(
                                asyncEscPosPrinter: AsyncEscPosPrinter,
                                codeException: Int
                            ) {
                                Log.e(
                                    "MainActivity",
                                    "AsyncEscPosPrint.OnPrintFinished : An error occurred !"
                                )
                            }

                            override fun onSuccess(asyncEscPosPrinter: AsyncEscPosPrinter) {
                                Log.i(
                                    "MainActivity",
                                    "AsyncEscPosPrint.OnPrintFinished : Print is finished !"
                                )
                            }
                        }
                    )
                        .execute(getAsyncEscPosPrinter(selectedDevice, position))
            }

        })
    }

    /**
     * Asynchronous printing
     */
    @SuppressLint("SimpleDateFormat")
    fun getAsyncEscPosPrinter(
        printerConnection: DeviceConnection?,
        position: Int
    ): AsyncEscPosPrinter {
        val format = SimpleDateFormat("'on' yyyy-MM-dd 'at' HH:mm:ss")
        val printer = AsyncEscPosPrinter(printerConnection, 203, 58f, 47)
        return printer.addTextToPrint(
            """
                [L]
                [L]<u>HUYEN TUI XACH</u>
                [L]
                [L]<u type='double'>${format.format(Date())}</u>
                [L]======================
                [L]<b>Comment:</b>${comments[position].comment}
                [L]<b>Nick name:</b>${comments[position].nickname}
                [L]<b>Unique Id:</b>${comments[position].uniqueId}
                [L]======================
                [L]<qrcode size='20'>https://www.tiktok.com/@${comments[position].uniqueId}</qrcode>
                [L]
                [L]
                
                """.trimIndent()
        )
    }

    companion object {
        fun hideKeyboard(activity: Activity) {
            val imm = activity.getSystemService(INPUT_METHOD_SERVICE) as InputMethodManager
            //Find the currently focused view, so we can grab the correct window token from it.
            var view = activity.currentFocus
            //If no view currently has focus, create a new one, just so we can grab a window token from it
            if (view == null) {
                view = View(activity)
            }
            imm.hideSoftInputFromWindow(view.windowToken, 0)
        }

        const val PERMISSION_BLUETOOTH = 1
        const val PERMISSION_BLUETOOTH_ADMIN = 2
        const val PERMISSION_BLUETOOTH_CONNECT = 3
        const val PERMISSION_BLUETOOTH_SCAN = 4
    }
}